data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,545 @@
+        Self(value.into())
+    }
+}
+
+impl Display for ConstString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{}", &**self)
+    }
+}
+
+impl Debug for ConstString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{:?}", &**self)
+    }
+}
+
+impl Hash for ConstString {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        (**self).hash(state)
+    }
+}
+
+impl PartialOrd for ConstString {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        PartialOrd::partial_cmp(&**self, &**other)
+    }
+}
+
+impl Ord for ConstString {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ord::cmp(&**self, &**other)
+    }
+}
+
+impl PartialEq for ConstString {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        PartialEq::eq(&**self, &**other)
+    }
+}
+
+macro_rules! impl_eq {
+    ($($ty:ty,)*) => {
+        impl_eq!($($ty),*);
+    };
+    ($($ty:ty),*) => {$(
+        impl PartialEq<$ty> for ConstString {
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &$ty) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+
+        impl PartialEq<ConstString> for $ty {
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &ConstString) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+    )*};
+}
+
+impl_eq!(String, str, &str);
+
+impl Eq for ConstString {}
+
+impl Serialize for ConstString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self)
+    }
+}
+
+impl Encode for ConstString {
+    fn size_hint(&self) -> usize {
+        self.as_bytes().size_hint()
+    }
+
+    fn encode_to<W: Output + ?Sized>(&self, dest: &mut W) {
+        self.as_bytes().encode_to(dest)
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        self.as_bytes().encode()
+    }
+
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        self.as_bytes().using_encoded(f)
+    }
+}
+
+impl IntoSchema for ConstString {
+    fn type_name() -> String {
+        String::type_name()
+    }
+    fn schema(map: &mut MetaMap) {
+        String::schema(map);
+    }
+}
+
+/// Union representing const-string variants: inlined or boxed.
+/// Distinction between variants are achieved by setting least significant bit for inlined variant.
+/// Boxed variant should have have 4/3/2 (depending on architecture 64/32/16 bit pointer width) trailing zeros due to pointer alignment.
+#[repr(C)]
+union ConstStringData {
+    inlined: InlinedString,
+    boxed: ManuallyDrop<BoxedString>,
+}
+
+impl ConstStringData {
+    #[inline]
+    const fn is_inlined(&self) -> bool {
+        self.inlined().is_inlined()
```

Comment:
this doesn't look correct

Reply:
When we check if the `ConstString` is inlined, we check the `len` field of the `InlinedString`, if the string is indeed inlined, then the last bit will be 1 and we will return true, if the string is actually boxed we will get the last byte of the pointer, the last bit of which should be 0 due to alignment in such a case, we return false.

Reply:
you're potentially creating a reference to an invalid object when calling inlined. This sounds like an UB

Reply:
if this is UB, now this tricky because you're going to have to use raw pointers instead of references

Reply:
I run tests with `miri` and it doesn't find any UB (i understand that `miri` can't find all UBs).
Also despite the fact that `InlinedString` could be in invalid state from logical point of view (wrong `len`, non-utf8 characters), it's seems to be correct from data point of view (one byte field and one array of bytes).

Also i don't understand how switching to pointers could help in that case.

Reply:
hm, this has me worried because I'm quite sure that fat to slim pointer casts that you've used are UB. How come miri didn't catch on that?

just to give you a feel of the problem, if you transmute `u8` (with a value other than 0 or 1) to `bool`, you're stepping into instant UB. This is because `bool` is a so-called non-robust type and has at least one trap representation. This may or may not be the case for strings, I'm not quite sure, I'll be looking it up some more

Reply:
you have to be thinking about what is allowed in the type system. Let's have a look at [rust reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#dangling-pointers) or [nomicon](https://doc.rust-lang.org/nomicon/what-unsafe-does.html) for a list of UBs. These things are UBs because the compiler is allowed to assume these types will never have invalid representations and make optimizations based on that assumption. 

Even if you're not accessing the string in your code, the compiler is free to inject operations that read from it as long as it produces output semantically equivalent to your source code. This is at least one reason why conversions like `12_u8` -> `bool` or `null_ptr` -> `&T` are instant UB, not when they are used/accessed in the source code. 

Another example is storing null pointer into `NonNull<T>` even though it is backed by `*mut T` which can be a null pointer. This is quite a similar situation to what you have in your code

How interleaved `std::String` is with the compiler is hard to say, but I'd be very cautious about producing invalid String representations, otherwise `from_utf8_unchecked` wouldn't be marked unsafe.